### PR TITLE
Extend the PDDL reader

### PR DESCRIPTION
**Pull request by alvalentini**

### CODE REVIEW

File: upf/io/pddl_reader.py [hunk 1]
```diff
@@ -14,24 +14,321 @@
...
+                self._add_timed_effects(problem, act, e)
+        elif len(eff) == 3 and op == 'at' and eff[1] == 'start':
+            self._add_effect(problem, act, eff[2], timing=upf.model.StartTiming())
+        elif len(eff) == 3 and op == 'at' and eff[1] == 'end':
+            self._add_effect(problem, act, eff[2], timing=upf.model.EndTiming())
+        else:
+            raise upf.exceptions.UPFProblemDefinitionError(f'Not able to handle: {eff}')
+
+
+    def parse_problem(self, domain_filename: str, problem_filename: str) -> 'upf.model.Problem':
```
Comment (mikand) [inline upf/io/pddl_reader.py:227]: An interesting feature could be to parse just the domain and get a partial `upf.model.Problem`: the use-case would be to have a fixed domain written in PDDL that gets instantiated with dynamic data. 

Would it be possible to make the `problem_filename` optional here?

File: upf/io/pddl_reader.py [hunk 2]
```diff
@@ -14,24 +14,321 @@
...
+        domain_res = pp_domain.parseFile(domain_filename)
+        problem_res = pp_problem.parseFile(problem_filename)
+
+        problem = upf.model.Problem(problem_res['name'], self._env,
+                                    initial_defaults={self._tm.BoolType(): self._em.FALSE()})
+
+        if 'predicates' in domain_res:
+            for p in domain_res['predicates']:
```
Comment (mikand) [inline upf/io/pddl_reader.py:235]:
```suggestion
        for p in domain_res.get('predicates', []):
```

File: upf/io/pddl_reader.py [hunk 3]
```diff
@@ -14,24 +14,321 @@
...
+            return self._em.Real(Fraction(exp))
+        else:
+            raise upf.exceptions.UPFProblemDefinitionError(f'Not able to handle: {exp}')
+
+    def _add_effect(self, problem, act, exp, cond=True, timing=None):
+        op = exp[0]
+        if op == 'and':
+            exp = exp[1:]
+            for e in exp:
+                self._add_effect(problem, act, e, cond, timing)
```
Comment (mikand) [inline upf/io/pddl_reader.py:156]: It should be very easy to remove recursion here: it suffices to implement a super-basic conjunctive partitioning like: https://github.com/pysmt/pysmt/blob/master/pysmt/rewritings.py#L856

File: upf/io/pddl_reader.py [hunk 4]
```diff
@@ -14,24 +14,321 @@
...
+            '-' : self._em.Minus,
+            '/' : self._em.Div,
+            '*' : self._em.Times
+        }
+
+    def _parse_exp(self, problem, act, var, exp):
+        if isinstance(exp, ParseResults) and len(exp) == 0:
+            return self._em.TRUE()
+        elif isinstance(exp, ParseResults) and exp[0] == '-' and len(exp) == 2:
+            return Times(-1, self._parse_exp(problem, act, var, exp[1]))
```
Comment (mikand) [inline upf/io/pddl_reader.py:117]: Maybe remove recursion here: it should be a matter of using a stack and a map for the results, no?

File: upf/io/pddl_reader.py [hunk 5]
```diff
@@ -14,24 +14,321 @@
...
+                        for j in range(1, len(dur)):
+                            v = Fraction(dur[j][2])
+                            if dur[j][0] == '>=':
+                                if lower is None or v > lower:
+                                    lower = v
+                            elif dur[j][0] == '<=':
+                                if upper is None or v < upper:
+                                    upper = v
+                            else:
+                                raise upf.exceptions.UPFProblemDefinitionError(f'Not able to handle duration constraint of action {n}')
```
Comment (mikand) [inline upf/io/pddl_reader.py:286]: Maybe use a `SyntaxError` instead?

File: upf/test/test_pddl_io.py [hunk 6]
```diff
@@ -234,6 +234,19 @@ def test_sailing_reader(self):
...
+        problem_filename = os.path.join(PDDL_DOMAINS_PATH, 'matchcellar', 'problem.pddl')
+        problem = reader.parse_problem(domain_filename, problem_filename)
+
+        self.assertTrue(problem is not None)
+        self.assertEqual(len(problem.fluents()), 4)
+        self.assertEqual(len(problem.actions()), 2)
+        self.assertEqual(len(problem.objects(problem.user_type('match'))), 3)
+        self.assertEqual(len(problem.objects(problem.user_type('fuse'))), 3)
+
     def test_matchcellar_writer(self):
```
Comment (mikand) [inline upf/test/test_pddl_io.py:250]: Ad a few more tests: e.g. for each problems in UPF format, print them in PDDL and read them back asserting that the actions and initial states and goals are the same modulo reordering

**Review by mikand — CHANGES_REQUESTED** [review]
@alvalentini Excellent work! I left some comments to be fixed, then we can merge!

File: upf/io/pddl_reader.py [hunk 7]
```diff
@@ -14,24 +14,321 @@
...
+            '=' : self._em.Equals,
+            '+' : self._em.Plus,
+            '-' : self._em.Minus,
+            '/' : self._em.Div,
+            '*' : self._em.Times
+        }
+
+    def _parse_exp(self, problem, act, var, exp):
+        if isinstance(exp, ParseResults) and len(exp) == 0:
+            return self._em.TRUE()
```
[the remark this answers is not in the document]
Comment (alvalentini) [reply upf/io/pddl_reader.py:115]: This happen only when we have something like `:precondition ()`

**Comment by mikand** [comment]
@alvalentini Coverage is low because now  the module `interop.tarski` becomes untested. Can you add a few tests to recover a bit of coverage? Or shall we wait for #72 to restore the coverage?

**Comment by alvalentini** [comment]
> @alvalentini Coverage is low because now the module `interop.tarski` becomes untested. Can you add a few tests to recover a bit of coverage? Or shall we wait for #72 to restore the coverage?

@mikand I think we can wait #72 that adds several tests about the `interop.tarski` module

**Pull request merged**